lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,565 @@
+#
+#
+# By default, ``FTSEarlyStopping`` and epoch-driven
+# transition criteria are composed. If a ``max_transition_epoch`` is specified for a given phase, the next finetuning phase will begin at that epoch unless ``FTSEarlyStopping`` criteria are met first.
+# If ``FinetuningScheduler.epoch_transitions_only`` is ``True``, ``FTSEarlyStopping`` will not be used
+# and transitions will be exclusively epoch-driven.
+#
+#
+# <div class="alert alert-info">
+#
+# **Tip:** Use of regex expressions can be convenient for specifying more complex schedules:
+#
+# <img src="emphasized_yaml.svg" style="width: 100%; height: auto; max-width:800px;">
+#
+# </div>
+#
+#
+#
+# The end-to-end example in this notebook ([Scheduled Finetuning For SuperGLUE](#superglue)) uses ``FinetuningScheduler`` in explicit mode to finetune a small foundational model on the [RTE](https://huggingface.co/datasets/viewer/?dataset=super_glue&config=rte) task of [SuperGLUE](https://super.gluebenchmark.com/).
+
+# %% [markdown]
+# ## Resuming Scheduled Finetuning Training Sessions
+#
+# Resumption of scheduled finetuning training is identical to the continuation of
+# [other training sessions](https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html) with the caveat that the provided checkpoint must have been saved by a ``FinetuningScheduler`` session.
+# ``FinetuningScheduler`` uses ``FTSCheckpoint`` (an extension of ``ModelCheckpoint``) to maintain schedule state with special metadata.
+#
+#
+# ```python
+# from pytorch_lightning import Trainer
+# from pytorch_lightning.callbacks.finetuning_scheduler import FinetuningScheduler
+# trainer = Trainer(callbacks=[FinetuningScheduler()], resume_from_checkpoint="some/path/to/my_checkpoint.ckpt")
+# ```
+#
+# Training will resume at the depth/level of the provided checkpoint according the specified schedule. Schedules can be altered between training sessions but schedule compatibility is left to the user for maximal flexibility. If executing a user-defined schedule, typically the same schedule should be provided for the original and resumed training sessions.
+#
+# By default (``FinetuningScheduler.restore_best`` is ``True``), ``FinetuningScheduler`` will attempt to restore the best available checkpoint before finetuning depth transitions.
+#
+# ```python
+# trainer = Trainer(
+#     callbacks=[FinetuningScheduler(new_incarnation_mode=True)],
+#     resume_from_checkpoint="some/path/to/my_kth_best_checkpoint.ckpt",
+# )
+# ```
+#
+# To handle the edge case wherein one is resuming scheduled finetuning from a non-best checkpoint and the previous best checkpoints may not be accessible, setting ``FinetuningScheduler.new_incarnation_mode`` to
+# ``True`` as above will re-intialize the checkpoint state with a new best checkpoint at the resumption depth.
+
+# %% [markdown]
+# <div class="alert alert-warning">
+#
+# **Note:** Currently, _FinetuningScheduler_ only supports the following _TrainingTypePlugins_:
```

Comment:
Note: This would be renamed Strategy for v1.7

Reply:
Good point, I've updated the reference here and will double-check to see if there are other appropriate references internal to finetuning_scheduler the module